### ISSUE TITLE
Deprecate jsonasobj2?

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#2217]
**Is your feature request related to a problem? Please describe.**

hello it is me the perf gremlin again.

when generating a pydantic module, effectively all the meaningful perf problems come from using jsonasobj2:

- within `induced_slot` 
  - (6.17% of total time) the `__bool__` method of jsonasobj
  - (12.85%) `__setattr__`
- within this dict comp: https://github.com/linkml/linkml/blob/bd49d48575d20e80875c2e1ef735d910f3a0d250/linkml/generators/pydanticgen/pydanticgen.py#L314
  - (21%) `as_dict`  

The rest is from parsing yaml, an inefficiency in schemaview's `__hash__` method (which i'm about to PR), and from black formatting. 

So perf is one thing, but also the `jsonasobj` root class in the generated python models give them a ~decent amount of surprising behavior (eg. [as i have noted before](https://github.com/linkml/linkml/issues/1665), which i now see [chris also noted previously](https://github.com/linkml/linkml/issues/971))

I am not quite sure what it buys us? the main thing seems to be using `obj.attr` instead of `obj['attr']` syntax? but that's also something we get with dataclasses, which are the main place they seem to be used.  

**Describe the solution you'd like**

I would like to just use dataclasses and dicts :).

I guess to make that happen (and to see if there is some functionality we want to preserve from `jsonasobj2`) i suppose i would need to have a better picture of why it is used/what features of it we like, but once i get that i'd be happy to PR.

**How important is this feature?** Select from the options below:
• Medium - strong perf impact which makes runtime use costly, hard to monkeypatch around.

**When will use cases depending on this become relevant?** Select from the options below:
• Short-term - 2-4 weeks

**Comment by sneakers-the-rat** [comment]
Just testing out what it would take, and i am getting almost all of the way there by doing this

https://github.com/sneakers-the-rat/linkml-runtime/commit/bc687a4bc8f2a6ee635fb6004d7882e2552c4953

and a handful of changes in `linkml` that are similar, usually calling `dataclasses.asdict` before some thing that expects it to be json-able.

anyway i don't think this would be hard and i would be willing to do it if we think this is a good idea

**Comment by cmungall** [comment]
I agree it would be good to remove dependence on jsonasobj2. You can see in this issue tracker a few confusing issues that have been caused by it over the years.

The only question is on strategy. The dependence on jsonasobj2 is quite woven in with the existing dataclass generator. We would want to be be very careful with changes - while most models are moving towards or have moved towards, a crucial one that has not is the metamodel.

Approaches

1. excise jsonasobj2 modifying as little as possible in the dataclass generator
2. doing a larger scale rewrite of the dataclass generator bring it in line with the patterns used in pydanticgen
3. move the metamodel to pydanticgen

I think 2 or 3 represent less work in the longer term, but 1 will yield perf gains sooner (and may be educational in its own right!)
